### ISSUE TITLE
Documentation related issues, questions and possible work items

### ISSUE THREAD

**Issue by chentong319** [onnx/onnx-mlir#1086]
This issue is opened to record issues, questions and possible work items when we update onnx-mlir documents.

1. Are we following docs/Documentation.md now? What's the role of docs/_data/navigation.yml?
2. What's our plan for the doc check tool?
3.  Merge of CONTRIBUTING.md and docs/Workflow.md
4. In the workflow, is it required to clone with ssh instead of https in order to create PR? Should we change the document accordingly?
5. In gen_onnx_mlir.py, there are several ways to generate `extraClassDeclaration`. Can be consolidated.
6. In gen_onnx_mlir.py, the customized builder may be avoided by using `returnType` in rewrite rules.
7. Restructure gen_onnx_mlir.py following PR#998.

**Comment by AlexandreEichenberger** [comment]
Answers for items for which I have an opinion

> 1. Are we following docs/Documentation.md now? What's the role of docs/_data/navigation.yml?

navigation is used for the onnx.ai/onnx-mlir website. This is what we have as the primary reference for our work.

> 2. What's our plan for the doc check tool?

check is to ensure that the script that we list in the md page is honest with the scripts. I am in favor of keeping it.

> 3. Merge of CONTRIBUTING.md and docs/Workflow.md. 

I suggest to merge info that is in Workflow into CONTRIBUTING. CONTRIBUTING is a very standard file to have in GitHub's projects.

> 4. In the workflow, is it required to clone with ssh instead of https in order to create PR? Should we change the document accordingly?

Personally, I think it should be required to work on a clone. I don't like it when folks push PRs directly on the main repo. SSH vs something else, I would refer to folks more expert in the field of security. I have no personal preference.

> 6. In gen_onnx_mlir.py, the customized builder may be avoided by using returnType in rewrite rules.

I have used custom builders and find this helpful to do very simple preprocessing. For example, when an op has a lot of parameters that are optional with default value, I like to have one with fewer parameters. They are trivial to write. Not familiar with the `returnType`'s approach.

**Comment by caoimhinuibrian** [comment]
> 4. In the workflow, is it required to clone with ssh instead of https in order to create PR? Should we change the document accordingly?

What exactly does this question mean? Why should we worry about how one clones the repo?

**Comment by caoimhinuibrian** [comment]
> Personally, I think it should be required to work on a clone. I don't like it when folks push PRs directly on the main repo. SSH vs something else, I would refer to folks more expert in the field of security. I have no personal preference.

@AlexandreEichenberger What do you mean by "push PRs directly to the main repo"?

**Comment by gongsu832** [comment]
> > 7\. In the workflow, is it required to clone with ssh instead of https in order to create PR? Should we change the document accordingly?
> 
> What exactly does this question mean? Why should we worry about how one clones the repo?

Indeed, https and ssh are just different ways to establish a secure connection to the github repo. They have no bearing on git operations such as pull, push, merge, etc.

**Comment by gongsu832** [comment]
> > Personally, I think it should be required to work on a clone. I don't like it when folks push PRs directly on the main repo. SSH vs something else, I would refer to folks more expert in the field of security. I have no personal preference.
> 
> @AlexandreEichenberger What do you mean by "push PRs directly to the main repo"?

I think he meant "work on a clone of a forked repo" instead of "work on a clone of the onnx-mlir repo", which I would agree.

Instead of cloning directly the original onnx-mlir repo and creating and pushing PRs on the onnx-mlir repo, one should create a fork of the original onnx-mlir repo and clone and work on the fork instead. We have privileged users working directly on the original onnx-mlir repo, which is not recommended.

**Comment by AlexandreEichenberger** [comment]
> Instead of cloning directly the original onnx-mlir repo and creating and pushing PRs on the onnx-mlir repo, one should create a fork of the original onnx-mlir repo and clone and work on the fork instead. We have privileged users working directly on the original onnx-mlir repo, which is not recommended.

yes, exactly
